### PR TITLE
Making Wink notification verification more robust.

### PR DESCRIPTION
**Pull request by danielrischar**
I'm trying to create a decent workflow for e2e testing of the wink subscription notifications.  In doing so I've run into a couple issue that can occur depending on server configuration/software for the callback end-point.

Fixes a potential issue with Wink notification verification caused by the payload contents being a string/object or vice versa. (#165)

Fixes a potential issue with reading HTTP headers as they are case-insensitive while their JSON representation is case-sensitive.

Adds HMACs to the error string for verification failure, useful for debugging purposes.

### CODE REVIEW

**Comment by danielrischar** [comment]
FYI, these scenarios are unit-testable so those are incoming.

File: org.opent2t.sample.hub.superpopular/com.wink.hub/js/thingTranslator.js [hunk 1]
```diff
@@ -10,6 +10,19 @@ var OpenT2T = require('opent2t').OpenT2T;
 var Crypto = require('crypto');
 
 /**
+ * Gets a property from a JSON dictionary without case sensitivity
+ * this will return the first item it finds, though multiple may match
+ * in JSON.
+ */
+function getDictionaryItemCaseInsensitive(obj, propertyName) {
+    for(var name in obj) {
+        if (propertyName.toUpperCase() == name.toUpperCase()) {
```
Comment (ChuckFerring) [inline org.opent2t.sample.hub.superpopular/com.wink.hub/js/thingTranslator.js]: >== [](start = 39, length = 2)

===?

Also, should you check if the object hasOwnProperty here?

File: org.opent2t.sample.hub.superpopular/com.wink.hub/js/thingTranslator.js [hunk 2]
```diff
@@ -45,16 +58,27 @@ class Translator {
         // Payload can contain one or more platforms defined using the provider schema.  This should return those platforms
         // converted to the opent2t/ocf representation.
         if (payload !== undefined) {
+
+            // The payload must be an object for translation, and a string/buffer for calculating
+            // the HMAC. Ensure we have a copy in both formats
+            var payloadAsString = typeof payload == 'object' ? JSON.stringify(payload) : payload;
+            var payloadAsObject = typeof payload == 'object' ? payload : JSON.parse(payload);
```
Comment (ChuckFerring) [inline org.opent2t.sample.hub.superpopular/com.wink.hub/js/thingTranslator.js]: ===?

File: org.opent2t.sample.hub.superpopular/com.wink.hub/js/thingTranslator.js [hunk 3]
```diff
@@ -45,16 +58,27 @@ class Translator {
...
-                var hashFromWink = verification.header["X-Hub-Signature"];
-                if (!this._verifyHmac(hashFromWink, verification.key, payload)) {
-                    throw new Error("Payload signature doesn't match.");
+                // HTTP headers are case insensitive, while the JSON dictionary is case sensitive,
+                // so the hub signature can be either x-hub-signature or X-Hub-Signature depending on
+                // what server was used to capture the notification POST request.
+                var hashFromWink = getDictionaryItemCaseInsensitive(verification.header, "x-hub-signature");
+                var hashFromPayload = this._generateHmac(verification.key, payloadAsString);
+
+                if (hashFromWink != hashFromPayload) {
```
Comment (ChuckFerring) [inline org.opent2t.sample.hub.superpopular/com.wink.hub/js/thingTranslator.js]: >!= [](start = 33, length = 2)

!==?

File: org.opent2t.sample.hub.superpopular/com.wink.hub/js/thingTranslator.js (hunk 1 above)
Comment (arjun-msft) [inline org.opent2t.sample.hub.superpopular/com.wink.hub/js/thingTranslator.js]: >propertyName.toUpperCase() [](start = 12, length = 26)

cache the propertyname toUpperCase once so you dont do it each time

Comment (danielrischar) [inline org.opent2t.sample.hub.superpopular/com.wink.hub/js/thingTranslator.js]: Can't use hasOwnProperty as it is case sensitive, I need to loop through all the properties and compare.

---
In reply to: [101857303](https://github.com/openT2T/translators/pull/166#discussion_r101857303) [](ancestors = 101857303)

**Comment by arjun-msft** [comment]
:shipit:

**Pull request merged**